src/models/AccessControlUnit.ts

```diff
@@ -0,0 +1,318 @@
+         * We don't want to invalidate those.
+         * @param map A map of entities to rules.
+         */
+        const removeRuleFromMap = (map: Map<string, ListRule[]>) => {
+            const entry = map.get(rule.entity);
+            if (entry !== undefined) {
+                const newEntry = entry.filter(internedRule => internedRule.sourceEvent.event_id !== rule.sourceEvent.event_id);
+                if (newEntry.length === 0) {
+                    map.delete(rule.entity);
+                } else {
+                    map.set(rule.entity, newEntry);
+                }
+            }
+        };
+        if (rule.isGlob()) {
+            removeRuleFromMap(this.globRules);
+        } else {
+            removeRuleFromMap(this.literalRules);
+        }
+    }
+
+    /**
+     * Add a rule to the cache e.g. it was added to a policy list.
+     * @param rule The rule to add.
+     */
+    private internRule(rule: ListRule) {
+        /**
+         * Add a rule to the map, there might be duplicates of this rule in other lists.
+         * @param map A map of entities to rules.
+         */
+        const addRuleToMap = (map: Map<string, ListRule[]>) => {
+            const entry = map.get(rule.entity);
+            if (entry !== undefined) {
+                entry.push(rule);
+            } else {
+                map.set(rule.entity, [rule]);
+            }
+        }
+        if (rule.isGlob()) {
+            addRuleToMap(this.globRules);
+        } else {
+            addRuleToMap(this.literalRules);
+        }
+    }
+
+    /**
+     * Update the cache for a single `ListRuleChange`.
+     * @param change The change made to a rule that was present in the policy list.
+     */
+    private updataCacheForChange(change: ListRuleChange): void {
+        if (change.rule.kind !== this.entityType || change.rule.recommendation !== this.recommendation) {
+            return;
+        }
+        switch (change.changeType) {
+            case ChangeType.Added:
+            case ChangeType.Modified:
+                this.internRule(change.rule);
+                break;
+            case ChangeType.Removed:
+                this.uninternRule(change.rule);
+                break;
+            default:
+                throw new TypeError(`Uknown ListRule change type: ${change.changeType}`);
+        }
+    }
+
+    /**
+     * Update the cache for a change in a policy list.
+     * @param changes The changes that were made to list rules since the last update to this policy list.
+     */
+    private updateCache(changes: ListRuleChange[]) {
+        changes.forEach(this.updataCacheForChange, this);
+    }
+}
+
+export enum Access {
+    /// The entity was explicitly banned by a policy list.
+    Banned,
+    /// The entity did not match any allow rule.
+    NotAllowed,
+    /// The user was allowed implicitly (it matched an allow rule and did not match any banning rules).
```

Comment:
The problem is there at least two meanings of explicitly allowed here, the main one being that anyone with a matching "allow" rule is explicitly allowed. Though they will not have access if they are also banned. If you make it through Allow, and Ban are you still explicitly allowed? It is also complicated further  if the only allow rule is "*"? . So it makes more sense to say you are only implicitly allowed by that point. It would be better if there was a way to avoid saying implicit completely though

Reply:
Which i will now do